packages/itwinui-react/src/core/ThemeProvider/ThemeProvider.tsx

```diff
@@ -211,29 +232,58 @@ const Root = React.forwardRef((props, forwardedRef) => {
 // ----------------------------------------------------------------------------
 
 /**
- * Returns theme from either parent context or by reading the closest
+ * Returns theme information from either parent ThemeContext or by reading the closest
  * data-iui-theme attribute if context is not found.
+ *
+ * Also returns the ThemeContext itself (if found).
  */
-const useParentTheme = () => {
+const useParentThemeAndContext = (rootElement: HTMLElement | null) => {
   const parentContext = React.useContext(ThemeContext);
-  const rootRef = React.useRef<HTMLElement>(null);
   const [parentThemeState, setParentTheme] = React.useState(
     parentContext?.theme,
   );
+  const [parentHighContrastState, setParentHighContrastState] = React.useState(
+    parentContext?.themeOptions?.highContrast,
+  );
+
+  const parentThemeRef = useLatestRef(parentContext?.theme);
 
   useIsomorphicLayoutEffect(() => {
-    setParentTheme(
-      (old) =>
-        old ||
-        (rootRef.current?.parentElement
-          ?.closest('[data-iui-theme]')
-          ?.getAttribute('data-iui-theme') as ThemeType),
+    // bail if we already have theme from context
+    if (parentThemeRef.current) {
+      return;
+    }
+
+    // find parent theme from closest data-iui-theme attribute
+    const closestRoot = rootElement?.parentElement?.closest('[data-iui-theme]');
+
+    if (!closestRoot) {
```

Comment:
nit: Is any comment needed here?

```suggestion
    // If no closestRoot exists, then there is nothing to listen to
    if (!closestRoot) {
```

Reply:
not necessary, the code is self explanatory